File: evalRPN.cpp
//TC: O(n)
//SC: O(n)

class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        
        if(tokens.empty())
            return 0;
        
        stack<int> s;
        int len = tokens.size();
        
        for(int i=0; i<len; i++){
            
            if(tokens[i] != "+" && tokens[i] != "-" && tokens[i] != "*" && tokens[i] != "/")
                s.push(stoi(tokens[i]));
            else{
                int op2 = s.top();
                s.pop();
                int op1 = s.top();
                s.pop();
                if(tokens[i] == "+")
                    s.push(op1 + op2);
                if(tokens[i] == "-")
                    s.push(op1 - op2);
                if(tokens[i] == "*")
                    s.push(op1 * op2);
                if(tokens[i] == "/")
                    s.push(op1/op2);
            }
            
        }
        
        return s.top();
        
        
    }
};

File: eval_RPN.py
# Time Complexity - O(n)
# Space Complexity - O(number of digits) for the stack

class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []  # initializing an empty stack

        for c in tokens:
            if c in "/+*-":  # checking if the top element is an operator
                b, a = stack.pop(), stack.pop()

                if c == '+':
                    stack.append(a + b)
                elif c == '-':
                    stack.append(a - b)
                elif c == '*':
                    stack.append(a * b)
                else:
                    stack.append(int(float(a) / b))
            else:  # if not an operator simply append to the stack
                stack.append(int(c))

        return stack.pop()
